python/cudf/cudf/tests/test_array_ufunc.py

```diff
@@ -70,35 +70,34 @@ def test_ufunc_index(ufunc):
     try:
         got = ufunc(*args)
     except AttributeError as e:
-        # We xfail if we don't have an explicit dispatch and cupy doesn't have
-        # the method so that we can easily identify these methods. As of this
+        # We check if an attribute error is raised so that if we don't
+        # have an explicit dispatch and cupy doesn't have the method
+        # so that we can easily identify these methods. As of this
         # writing, the only missing methods are isnat and heaviside.
         if "module 'cupy' has no attribute" in str(e):
-            pytest.xfail(reason="Operation not supported by cupy")
-        raise
+            with pytest.raises(AttributeError):
+                raise e
+            return
+        else:
+            raise
```

Comment:
This is a pattern I have to introduce in a bunch of places, which I don't _really_ like, but it's better than xfail which is kind of "wrong"?

Reply:
Would it help to introduce something like the `expect_warning_if` contextmanager that I just created, except in this case for exception handling? https://github.com/rapidsai/cudf/blob/branch-23.02/python/cudf/cudf/testing/_utils.py#L393

Reply:
That may also help reduce or usage of the `nullcontext`.

Reply:
In the end, I went with the suggestion below of using `request.applymarker` since that feels a bit more ergonomic.

Reply:
Sounds good. That works much better for failures than warnings.